pkg/cmd/list.go

```diff
@@ -119,67 +129,67 @@ func NewWhoCanOptions(configFlags *clioptions.ConfigFlags,
-	if err != nil {
-		return nil, fmt.Errorf("getting config: %v", err)
-	}
-
-	client, err := kubernetes.NewForConfig(clientConfig)
-	if err != nil {
-		return nil, fmt.Errorf("creating client: %v", err)
-	}
-
-	mapper, err := configFlags.ToRESTMapper()
-	if err != nil {
-		return nil, fmt.Errorf("getting mapper: %v", err)
-	}
-
-	clientNamespace := client.CoreV1().Namespaces()
-	accessChecker := NewAccessChecker(client.AuthorizationV1().SelfSubjectAccessReviews())
-	namespaceValidator := NewNamespaceValidator(clientNamespace)
-	resourceResolver := NewResourceResolver(client.Discovery(), mapper)
-
-	o := NewWhoCanOptions(configFlags,
-		configFlags.ToRawKubeConfigLoader(),
-		clientNamespace,
-		client.RbacV1(),
-		namespaceValidator,
-		resourceResolver,
-		accessChecker,
-		NewPolicyRuleMatcher(),
-		streams)
+	o := newWhoCanOptions()
 
 	cmd := &cobra.Command{
 		Use:          whoCanUsage,
 		Long:         whoCanLong,
 		Example:      whoCanExample,
 		SilenceUsage: true,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if err := o.Complete(args); err != nil {
+			clientConfig, err := o.configFlags.ToRESTConfig()
+			if err != nil {
+				return fmt.Errorf("getting config: %v", err)
+			}
+
+			client, err := kubernetes.NewForConfig(clientConfig)
+			if err != nil {
+				return fmt.Errorf("creating client: %v", err)
+			}
+
+			mapper, err := o.configFlags.ToRESTMapper()
+			if err != nil {
+				return fmt.Errorf("getting mapper: %v", err)
+			}
+
+			clientNamespace := client.CoreV1().Namespaces()
+			accessChecker := NewAccessChecker(client.AuthorizationV1().SelfSubjectAccessReviews())
+			namespaceValidator := NewNamespaceValidator(clientNamespace)
+			resourceResolver := NewResourceResolver(client.Discovery(), mapper)
+
+			wc := NewWhoCan(o,
+				o.configFlags.ToRawKubeConfigLoader(),
+				clientNamespace,
+				client.RbacV1(),
+				namespaceValidator,
+				resourceResolver,
+				accessChecker,
+				NewPolicyRuleMatcher(),
+				streams)
+
+			if err := wc.Complete(args); err != nil {
 				return err
 			}
-			if err := o.Validate(); err != nil {
+			if err := wc.Validate(); err != nil {
 				return err
 			}
-			if err := o.Check(); err != nil {
+			if err := wc.Check(); err != nil {
 				return err
 			}
 
 			return nil
 		},
 	}
 
-	cmd.PersistentFlags().StringVar(&o.subResource, "subresource", o.subResource,
+	cmd.Flags().StringVar(&o.subResource, "subresource", o.subResource,
 		"SubResource such as pod/log or deployment/scale")
-	cmd.PersistentFlags().BoolVarP(&o.allNamespaces, "all-namespaces", "A", false,
+	cmd.Flags().BoolVarP(&o.allNamespaces, "all-namespaces", "A", o.allNamespaces,
 		"If true, check the specified action in all namespaces.")
 
-	flag.CommandLine.VisitAll(func(goflag *flag.Flag) {
-		cmd.PersistentFlags().AddGoFlag(goflag)
+	flag.CommandLine.VisitAll(func(gf *flag.Flag) {
+		cmd.Flags().AddGoFlag(gf)
 	})
-	configFlags.AddFlags(cmd.Flags())
+	o.configFlags.AddFlags(cmd.Flags())
```

Comment:
I wonder if a simpler fix would have just been to move the lines that were at lines 124-152 to be after these flags have been set up? I don't love the extra complexity of having the separate `whoCanOptions` struct just to hold the `configFlags` (but am I missing something?)